internal/catalogmetadata/client/client.go

```diff
@@ -43,66 +47,93 @@ type Client struct {
 	fetcher Fetcher
 }
 
-func (c *Client) Bundles(ctx context.Context) ([]*catalogmetadata.Bundle, error) {
+func (c *Client) Bundles(ctx context.Context, packageName string) ([]*catalogmetadata.Bundle, error) {
 	var allBundles []*catalogmetadata.Bundle
 
 	var catalogList catalogd.ClusterCatalogList
 	if err := c.cl.List(ctx, &catalogList); err != nil {
 		return nil, err
 	}
+
+	var errs []error
 	for _, catalog := range catalogList.Items {
-		// if the catalog has not been successfully unpacked, skip it
+		// if the catalog has not been successfully unpacked, report an error. This ensures that our
+		// reconciles are deterministic and wait for all desired catalogs to be ready.
 		if !meta.IsStatusConditionPresentAndEqual(catalog.Status.Conditions, catalogd.TypeUnpacked, metav1.ConditionTrue) {
+			errs = append(errs, fmt.Errorf("catalog %q is not unpacked", catalog.Name))
 			continue
 		}
 		channels := []*catalogmetadata.Channel{}
 		bundles := []*catalogmetadata.Bundle{}
 		deprecations := []*catalogmetadata.Deprecation{}
 
-		rc, err := c.fetcher.FetchCatalogContents(ctx, catalog.DeepCopy())
+		catalogFS, err := c.fetcher.FetchCatalogContents(ctx, catalog.DeepCopy())
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error fetching catalog %q contents: %v", catalog.Name, err))
+			continue
+		}
+
+		packageFS, err := fs.Sub(catalogFS, packageName)
 		if err != nil {
-			return nil, fmt.Errorf("error fetching catalog contents: %s", err)
+			errs = append(errs, fmt.Errorf("error reading package subdirectory %q from catalog %q filesystem: %v", packageName, catalog.Name, err))
+			continue
 		}
-		defer rc.Close()
 
-		err = declcfg.WalkMetasReader(rc, func(meta *declcfg.Meta, err error) error {
+		var (
+			channelsMu     sync.Mutex
+			bundlesMu      sync.Mutex
+			deprecationsMu sync.Mutex
+		)
+
+		if err := declcfg.WalkMetasFS(ctx, packageFS, func(_ string, meta *declcfg.Meta, err error) error {
```

Comment:
Is this function now guaranteed to run concurrently? Is there any guarantees on the same `*declcfg.Meta` object not being processed more than once?

Reply:
It runs concurrently by default (based on number of CPUs Go thinks are available). There is an exactly once guarantee for an individual `*declcfg.Meta` object.

Reply:
Ultimately, we're parsing `n` files at a time (where `n` is number of goroutines). You may notice that the cache write metas into their own files. This is on purpose to make as many files as possible so that as many goroutines as possible (up to `n`, at least) have work to do.